lib/logstash/codecs/cloudwatch_logs.rb

```diff
@@ -25,11 +31,13 @@ def decompress(data)
   end
 
   def parse(json, &block)
-    base = json.reject { |k,_| k == "logEvents" }.freeze
-    events = json["logEvents"]
+    events = json.delete("logEvents")
+    json.freeze
 
     events.each do |event|
-      yield LogStash::Event.new(base.merge(event))
+      epochmillis = event.delete("timestamp").to_i
+      event[LogStash::Event::TIMESTAMP] = LogStash::Timestamp.at(epochmillis / 1000, (epochmillis % 1000) * 1000)
```

Comment:
The 'timestamp' field of an event is supposed to indicate the time that the event took place, as determined by the event source. This is distinct from the event ingest timestamp, which may be off by several minutes or even be completely out of order, due to client-side buffering.

If you are using the Logs agent on a host, it will either set the timestamp based on when the event was generated on the client, or parse it out of the event itself prior to calling the PutLogEvents function. If your log events contain a parseable timestamp that does not match the event's timestamp field, then you are "doing it wrong".

I would suggest that the plugin set the Logstash timestamp field by default based on the Logs event timestamp; if for some reason people are misusing Logs and not setting the timestamp properly they can still use the date filter to override it by parsing the message field - but they really should be doing this at the time the events are being stored into Logs in order to get the full value out of the product.